### PR TITLE
[buildkite] Implement secrets encryption and sandboxing

### PR DESCRIPTION
**Pull request by staticfloat**
This adds a proof-of-concept demonstration of two new buildkite plugins:

* `cryptic` adds secrets management to privileged pipelines.  These
pipelines cannot be freely modified; their integrity is verified against
a signature maintained by committers with a secret key.  This allows
certain portions of the CI configuration (which are privileged and can
decrypt encrypted files/environment variables) to remain public, but
read-only to the general populace.

*  `sandbox` adds a generic sandboxing mechanism that allows CI steps to
be run within user-provided rootfs images.  We're using these here to
provide compiler toolchains for the `llvm-passes` CI steps, and the plan
is to eventually provide _all_ compiler toolchains through such rootfs
images.

### CODE REVIEW

File: .buildkite/llvm_passes.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      make -j 6 -C deps install-llvm install-clang install-llvm-tools install-libuv install-utf8proc install-unwind
+      echo "+++ run clangsa/analyzegc"
+      make -j 6 -C test/clangsa
+      make -j 6 -C src analyzegc
+    timeout_in_minutes: 60
+
+  - label: "llvmpasses"
+    plugins:
+      - staticfloat/sandbox#v1:
+          rootfs_url: https://github.com/JuliaCI/rootfs-images/releases/download/v1/llvm-passes.tar.gz
```
Comment (vchuravy) [inline .buildkite/llvm_passes.yml:31]: Does this go through our CDN? Can we cache these?

File: .buildkite/rootfs_images/README.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+## Rootfs images
+
+Our CI setup makes use of rootfs images that contain our build tools.
+These rootfs images are built using the fairly simple
```
Comment (vchuravy) [inline .buildkite/rootfs_images/README.md:4]: incomplete?

File: .buildkite/llvm_passes.yml (hunk 1 above)
Comment (staticfloat) [reply .buildkite/llvm_passes.yml:31]: Yes, it should automatically get cached.  It gets downloaded as an artifact, and all artifacts are automatically cached as part of the `JuliaCI/julia` plugin.

**Comment by staticfloat** [comment]
giphy

🎉

**Comment by staticfloat** [comment]
@maleadt you're probably interested in this as well.

The workers are running via the configurations curated here: https://github.com/JuliaCI/sandboxed-buildkite-agent/blob/main/linux-sandbox.jl/build_systemd_config.jl

I intend to add windows/macos/freebsd setups there as well in the future (these workers will only be for base CI, I don't intend to run CI for the ecosystem on these).  I am using the sandbox as the isolation mechanism for these workers as well because it's nestable.  This allows us to run sandboxes inside of sandboxes, which is how we're able to do this rootfs nonsense for the LLVM pipelines.

Eventually, our buildbot configuration is just going to be a rootfs image that gets downloaded onto a machine and unpacked as an artifact.  And yes, this is basically just us reinventing Docker, but since they can't nest, it's just impossible to use it securely without opening the host up for trivial exploitation.

Also, this showcases some simple secrets management with `cryptic`.  I'll expand the example here after this gets merged and I expand Dilum's codecov work.

**Comment by DilumAluthge** [comment]
I disabled Buildkite commit statuses since they were failing on some PRs (e.g. https://github.com/JuliaLang/julia/pull/41264).

So we should remember to re-enable Buildkite commit statuses right before we merge this PR.

**Comment by staticfloat** [comment]
I have re-enabled buildkite commit statuses

**Comment by DilumAluthge** [comment]
Does this need to be backported to 1.6 and 1.7? Otherwise, Buildkite statuses will always be failing on the `release-1.6` and `release-1.7` branches.

**Comment by staticfloat** [comment]
Sure, it can be backported.

**Pull request merged**